src/uart.c

```diff
@@ -0,0 +1,67 @@
+#include "uart.h"
+#include "utils.h"
+#include "gpio.h"
+
+void    uart_configure(UART_config *config)
+{
+    // TODO: add all UART ports
+    if (config->port == UART1)
+    {
+        gpio_select_alternate_function(GPIOA, GPIO_PIN_8, AF1);
+        gpio_select_alternate_function(GPIOA, GPIO_PIN_9, AF1);
+        gpio_select_alternate_function(GPIOA, GPIO_PIN_10, AF1);
+    }
+
+    switch (config->word_length)
+    {
+        case EIGHT_BITS:
+            config->port->CR1 = set_bits_with_offset(config->port->CR1, M1_OFFSET, 1, 0);
```

Comment:
M[1:0] can only be written when UART is disabled. This is unlikely to cause issue now, but it's safer to set UE=0 before doing the setup.

Reply:
Also, try to set both M1 and M0 in one go.